### PR TITLE
Add retry to purge_inactive_dag_warnings

### PR DESCRIPTION
**Pull request by michaelmicheal**
This PR safely catches exceptions to the `purge_inactive_dag_warnings`. This prevents unhandled exceptions from causing the DagFileProcessorManager to fail.
related: #28480 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/models/dagwarning.py [hunk 1]
```diff
@@ -79,7 +83,12 @@ def purge_inactive_dag_warnings(cls, session: Session = NEW_SESSION) -> None:
             query = session.query(cls).filter(cls.dag_id.in_(dag_ids))
         else:
             query = session.query(cls).filter(cls.dag_id == DagModel.dag_id, DagModel.is_active == false())
-        query.delete(synchronize_session=False)
+        try:
+            query.delete(synchronize_session=False)
+        except OperationalError:
+            # If the purge query fails, it's not critical to the dag processor
+            # We ignore it so the dag processor manager doesn't exit
+            log.error("Failed to purge inactive dag_warnings, ignoring")
```
Comment (kaxil) [inline airflow/models/dagwarning.py]: You can instead re-try the transaction https://github.com/apache/airflow/blob/23264fb820c179e9951ea9706f68b13a9b3fdbc0/airflow/utils/retries.py#L49

Comment (michaelmicheal) [reply airflow/models/dagwarning.py]: What do you think of this? Also, do you think we need to run this query every parsing loop? What if we only run it every x seconds and make that configurable?

**Review by ephraimbuddy — APPROVED** [review]

File: airflow/models/dagwarning.py [hunk 2]
```diff
@@ -72,6 +73,11 @@ def purge_inactive_dag_warnings(cls, session: Session = NEW_SESSION) -> None:
 
         :return: None
         """
+        cls._purge_inactive_dag_warnings_with_retry(session)
+
+    @classmethod
+    @retry_db_transaction
+    def _purge_inactive_dag_warnings_with_retry(cls, session: Session) -> None:
```
Comment (ephraimbuddy) [inline airflow/models/dagwarning.py:80]: LGTM, let's add a little test

File: tests/dag_warning/test_dag_warning.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+import unittest
+from unittest.mock import MagicMock
+
+from sqlalchemy.exc import OperationalError
+
+from airflow.models.dagwarning import DagWarning
+
+
+class TestDagWarning(unittest.TestCase):
```
Comment (uranusjr) [inline tests/dag_warning/test_dag_warning.py]: By convention, this should go into `tests/models/test_dagwarning.py` instead (matching the source tree’s structure).

File: tests/models/test_dagwarning.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+import unittest
+from unittest.mock import MagicMock
+
+from sqlalchemy.exc import OperationalError
+
+from airflow.models.dagwarning import DagWarning
+
+
+class TestDagWarning(unittest.TestCase):
```
Comment (ephraimbuddy) [inline tests/models/test_dagwarning.py]: We now use pytest for our tests. With pytest, you have access to `session` fixture

Comment (michaelmicheal) [reply tests/models/test_dagwarning.py]: Are there any good examples of using the shared session fixture?

Comment (ephraimbuddy) [reply tests/models/test_dagwarning.py]: Check this: https://github.com/apache/airflow/blob/378c8faba19d21b29abfff1705a8e7624aeabdc4/tests/models/test_taskinstance.py#L507

Comment (michaelmicheal) [reply tests/models/test_dagwarning.py]: @ephraimbuddy What do you think of the tests?

File: tests/models/test_dagwarning.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        clear_db_dags()
+
+    def test_purge_inactive_dag_warnings(self, session):
+        """
+        Test that the purge_inactive_dag_warnings method deletes inactive dag warnings
+        """
+
+        dags = [DagModel(dag_id="dag_1", is_active=False), DagModel(dag_id="dag_2", is_active=True)]
+        for dag in dags:
+            session.add(dag)
```
Comment (ephraimbuddy) [inline tests/models/test_dagwarning.py]:
```suggestion
        session.add_all(dags)
```

File: tests/models/test_dagwarning.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        session.commit()
+
+        dag_warnings = [
+            DagWarning("dag_1", "non-existent pool", "non-existent pool"),
+            DagWarning("dag_2", "non-existent pool", "non-existent pool"),
+        ]
+        for dag_warning in dag_warnings:
+            session.add(dag_warning)
```
Comment (ephraimbuddy) [inline tests/models/test_dagwarning.py]:
```suggestion
        session.add_all(dag_warnings)
```

File: tests/models/test_dagwarning.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        self.session_mock = MagicMock()
+        self.delete_mock = MagicMock()
+        self.session_mock.query.return_value.filter.return_value.delete = self.delete_mock
+
+        self.delete_mock.side_effect = [OperationalError(None, None, "database timeout"), None]
+
+        DagWarning.purge_inactive_dag_warnings(self.session_mock)
+
+        # Assert that the delete method was called twice
+        assert self.delete_mock.call_count == 2
```
Comment (ephraimbuddy) [inline tests/models/test_dagwarning.py:69]: Why only twice? What's the default number of DB retries?

**Review by ephraimbuddy — APPROVED** [review]
LGTM, just nits

File: tests/models/test_dagwarning.py (hunk 7 above)
Comment (michaelmicheal) [reply tests/models/test_dagwarning.py:69]: This tests that if it fails once and then works, we do in fact stop after the successful run. How much do I need to test that the retry does in fact work?

**Comment by michaelmicheal** [comment]
@uranusjr, @kaxil , is this good to merge?

**Review by uranusjr — APPROVED** [review]

**Comment by kaxil** [comment]
🎉

**Pull request merged**